File: app/src/main/java/xyz/sleekstats/completist/view/MovieAdapter.java
package xyz.sleekstats.completist.view;

import android.support.annotation.NonNull;
import android.support.v7.widget.CardView;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.List;

import xyz.sleekstats.completist.R;
import xyz.sleekstats.completist.databinding.FilmItemBinding;
import xyz.sleekstats.completist.model.FilmByPerson;

//Load film names and posters for a specific actor/director in a recyclerview
public class MovieAdapter extends RecyclerView.Adapter<MovieAdapter.FilmViewHolder> {

    private List<FilmByPerson> mCurrentMovieList;
    private ItemClickListener mClickListener;

    public MovieAdapter(List<FilmByPerson> filmByPersonList) {
        this.mCurrentMovieList = filmByPersonList;
    }

    @NonNull
    @Override
    public FilmViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater layoutInflater = LayoutInflater.from(parent.getContext());
        FilmItemBinding filmItemBinding = FilmItemBinding.inflate(layoutInflater, parent, false);
        return new FilmViewHolder(filmItemBinding);
    }

    @Override
    public void onBindViewHolder(@NonNull FilmViewHolder holder, int position) {
        FilmByPerson movie = mCurrentMovieList.get(position);
        holder.bind(movie);
        holder.setPos(position);
    }

    public void setCurrentMovieList(List<FilmByPerson> mCurrentMovieList) {
        this.mCurrentMovieList = mCurrentMovieList;
        notifyDataSetChanged();
    }

    @Override
    public int getItemCount() {
        return mCurrentMovieList.size();
    }

    class FilmViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        private final CardView mView;
        private int listPos;
        private FilmByPerson mFilm;
        private final FilmItemBinding filmItemBinding;

        public FilmViewHolder(FilmItemBinding binding) {
            super(binding.getRoot());
            this.filmItemBinding = binding;
            mView = (CardView) itemView;
            itemView.findViewById(R.id.title).setOnClickListener(this);
            itemView.findViewById(R.id.poster).setOnClickListener(this);
            itemView.findViewById(R.id.watched_btn).setOnClickListener(this);
            itemView.findViewById(R.id.later_btn).setOnClickListener(this);
            itemView.findViewById(R.id.ignore_btn).setOnClickListener(this);
        }

        public void bind(FilmByPerson film) {
            String id = film.getId();
            mView.setTag(id);
            mFilm = film;
            filmItemBinding.setFilm(film);
            filmItemBinding.executePendingBindings();
        }

        public void setPos(int pos) {
            listPos = pos;
        }


        @Override
        public void onClick(View view) {
            if (mClickListener == null) {
                return;
            }
            String id = String.valueOf(mView.getTag());
            int viewId = view.getId();
            switch (viewId) {
                case R.id.title:
                case R.id.poster:
                    mClickListener.onFilmClick(id);
                    break;
                case R.id.watched_btn:
                    mClickListener.onFilmWatched(listPos);
                    break;
                case R.id.later_btn:
                    mClickListener.onFilmQueued(listPos);
                    break;
                case R.id.ignore_btn:
                    break;
            }
        }
    }

    public void setClickListener(ItemClickListener itemClickListener) {
        this.mClickListener = itemClickListener;
    }

    public interface ItemClickListener {
        void onFilmClick(String movieID);
        void onFilmWatched(int pos);
        void onFilmQueued(int pos);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }
}

File: app/src/main/java/xyz/sleekstats/completist/view/MovieAdapter.kt
package xyz.sleekstats.completist.view

import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.cardview.widget.CardView
import androidx.recyclerview.widget.RecyclerView
import xyz.sleekstats.completist.R
import xyz.sleekstats.completist.databinding.FilmItemBinding
import xyz.sleekstats.completist.model.MediaByPerson
import xyz.sleekstats.completist.view.MovieAdapter.FilmViewHolder

//Load film names and posters for a specific actor/director in a recyclerview
class MovieAdapter(private var mCurrentMovieList: List<MediaByPerson>) : RecyclerView.Adapter<FilmViewHolder>() {

    private var mClickListener: ItemClickListener? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): FilmViewHolder =
            FilmViewHolder(FilmItemBinding.inflate(LayoutInflater.from(parent.context), parent, false))

    override fun onBindViewHolder(holder: FilmViewHolder, position: Int) {
        holder.apply {
            bind(mCurrentMovieList[position])
            setPos(position)
        }
    }

    override fun getItemCount(): Int = mCurrentMovieList.size

    inner class FilmViewHolder(private val filmItemBinding: FilmItemBinding) : RecyclerView.ViewHolder(filmItemBinding.root), View.OnClickListener {
        private val mView: CardView = itemView as CardView
        private var listPos = 0
        private var mFilm: MediaByPerson? = null

        init {
            itemView.findViewById<View>(R.id.title).setOnClickListener(this)
            itemView.findViewById<View>(R.id.poster).setOnClickListener(this)
            itemView.findViewById<View>(R.id.watched_btn).setOnClickListener(this)
            itemView.findViewById<View>(R.id.later_btn).setOnClickListener(this)
        }

        fun bind(film: MediaByPerson) {
            val id = film.id
            mView.tag = id
            mFilm = film
            filmItemBinding.film = film
            filmItemBinding.executePendingBindings()
        }

        fun setPos(pos: Int) {
            listPos = pos
        }

        override fun onClick(view: View) {
            mClickListener?.let {
                val id = mView.tag.toString()
                when (view.id) {
                    R.id.title, R.id.poster -> it.onFilmClick(id, mFilm?.isFilm == true)
                    R.id.watched_btn -> it.onFilmWatched(listPos)
                    R.id.later_btn -> it.onFilmQueued(listPos)
                }
            }
        }
    }

    fun setClickListener(itemClickListener: ItemClickListener?) {
        mClickListener = itemClickListener
    }

    override fun getItemId(position: Int): Long = position.toLong()

    interface ItemClickListener {
        fun onFilmClick(movieID: String?, isFilm: Boolean)
        fun onFilmWatched(pos: Int)
        fun onFilmQueued(pos: Int)
    }

}